lib/fs/memfs/memfs.c

```diff
@@ -253,6 +253,42 @@ static ssize_t memfs_read(filecookie *fcookie, void *buf, off_t off, size_t len)
     return len;
 }
 
+static status_t memfs_truncate(filecookie *fcookie, uint64_t len)
+{
+    LTRACEF("filecookie %p, len %llu\n", fcookie, len);
+
+    status_t rc = NO_ERROR;
+
+    memfs_file_t *file = (memfs_file_t *)fcookie;
+
+    mutex_acquire(&file->fs->lock);
+
+    // Can't use truncate to grow a file.
+    if (len > file->len)
+        return ERR_INVALID_ARGS;
+
+    void *ptr = realloc(file->ptr, len);
+    if (unlikely(ptr == NULL)) {
+        rc = ERR_NO_MEMORY;
+        goto finish;
+    }
+
+    // Fast-path: It's very likely that realloc won't move the allocated memory
+    // block since we're shrinking an existing file. If the block hasn't been
+    // moved, we can terminate early.
+    if (ptr == file->ptr) {
+        rc = NO_ERROR;
+        goto finish;
+    }
+
+    // If we get a different pointer back, we need to move the file.
+    file->ptr = memmove(ptr, file->ptr, len);
```

Comment:
That's not how realloc works. It actually does the memmove for you and frees the old block (if that's how it chooses to do it). So, if you get a valid pointer back from realloc it's okay to use. The only real edge case is if it returns a null when you didn't expect it and thus you have to restore the old pointer.